File: src/main/java/uk/gov/justice/hmpps/casenotes/config/SecurityUserContext.java
package uk.gov.justice.hmpps.casenotes.config;

import org.apache.commons.lang3.RegExUtils;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Component
public class SecurityUserContext {

	private Authentication getAuthentication() {
		return SecurityContextHolder.getContext().getAuthentication();
	}

	public String getCurrentUsername() {
		String username;

		Object userPrincipal = getUserPrincipal();

		if (userPrincipal instanceof String) {
			username = (String) userPrincipal;
		} else if (userPrincipal instanceof UserDetails) {
			username = ((UserDetails)userPrincipal).getUsername();
		} else if (userPrincipal instanceof Map) {
			Map userPrincipalMap = (Map) userPrincipal;
			username = (String) userPrincipalMap.get("username");
		} else {
			username = null;
		}

		return username;
	}

	private Object getUserPrincipal() {
		Object userPrincipal = null;

		final Authentication auth = getAuthentication();

		if (auth != null) {
			userPrincipal = auth.getPrincipal();
		}
		return userPrincipal;
	}

	public static boolean hasRoles(final String... allowedRoles) {
		final var roles = Arrays.stream(allowedRoles)
				.map(r -> RegExUtils.replaceFirst(r, "ROLE_", ""))
				.collect(Collectors.toList());

		return hasMatchingRole(roles, SecurityContextHolder.getContext().getAuthentication());
	}

	public boolean isOverrideRole(final String... overrideRoles) {
		final var roles = Arrays.asList(overrideRoles.length > 0 ? overrideRoles : new String[]{"SYSTEM_USER"});
		return hasMatchingRole(roles, getAuthentication());
	}

	private static boolean hasMatchingRole(final List<String> roles, final Authentication authentication) {
		return authentication != null &&
				authentication.getAuthorities().stream()
						.anyMatch(a -> roles.contains(RegExUtils.replaceFirst(a.getAuthority(), "ROLE_", "")));
	}
}

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/SecurityUserContext.kt
package uk.gov.justice.hmpps.casenotes.config

import lombok.extern.slf4j.Slf4j
import org.apache.commons.lang3.RegExUtils
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component

@Slf4j
@Component
class SecurityUserContext {

  fun getAuthentication(): Authentication = SecurityContextHolder.getContext().authentication

  fun getCurrentUsername(): String? = getAuthorisedUser()?.username

  fun getCurrentUser(): UserIdUser = getAuthorisedUser()
    ?: throw IllegalStateException("Current user not set but is required")

  fun hasAnyRole(vararg roles: String): Boolean = hasMatchingRole(roles.toList())

  fun isOverrideRole(vararg overrideRoles: String): Boolean =
    hasMatchingRole(if (overrideRoles.isEmpty()) listOf("SYSTEM_USER") else overrideRoles.toList())

  private fun getAuthorisedUser(): UserIdUser? =
    when (val auth = getAuthentication()) {
      is AuthAwareAuthenticationToken -> auth.userIdUser
      else -> null
    }

  private fun hasMatchingRole(roles: List<String>): Boolean =
    getAuthentication().authorities?.any { a: GrantedAuthority? ->
      roles.contains(RegExUtils.replaceFirst(a!!.authority, "ROLE_", ""))
    } == true

  data class UserIdUser(val username: String, val userId: String)

  companion object {
    const val ROLE_CASE_NOTES_READ = "ROLE_PRISONER_CASE_NOTES__RO"
    const val ROLE_CASE_NOTES_WRITE = "ROLE_PRISONER_CASE_NOTES__RW"
    const val ROLE_CASE_NOTES_SYNC = "ROLE_PRISONER_CASE_NOTES__SYNC__RW"
    const val ROLE_SYSTEM_GENERATED_RW = "ROLE_PRISONER_CASE_NOTES__SYSTEM_GENERATED__RW"
  }
}
